apps/script/HoC2023AiGenerateWeights.py

```diff
@@ -107,33 +111,83 @@
 background_dict = {}
 foreground_dict = {}
 
-# Calculate and print similarity scores
-for emoji_name in emoji_names_list:
-    palette_scores = []
-    for palette_word in palette_model_descriptive_names:
-        id_token = nlp(emoji_name)
-        palette_token = nlp(palette_word)
-        similarity_score = id_token.similarity(palette_token)
-        palette_scores.append(round(similarity_score, 2))
+# Define caches to store ada's raw embedding outputs to reduce query costs
+# These caches are stored as pickle files; python's native way to serialize data
+embedding_paths = ['apps/static/dance/ai/model/input_embeddings.pkl',
+'apps/static/dance/ai/model/palette_embeddings.pkl',
+'apps/static/dance/ai/model/background_embeddings.pkl',
+'apps/static/dance/ai/model/foreground_embeddings.pkl']
+input_path = embedding_paths[0]
+palette_path = embedding_paths[1]
+background_path = embedding_paths[2]
+foreground_path = embedding_paths[3]
+
+def load_embedding_cache(path):
+    # load the cache if it exists, and save a copy to disk
+    try:
+        embedding_cache = pd.read_pickle(path)
+    except FileNotFoundError:
+        embedding_cache = {}
+    with open(path, "wb") as embedding_cache_file:
+        pickle.dump(embedding_cache, embedding_cache_file)
+    return embedding_cache
+
+caches = [load_embedding_cache(path) for path in embedding_paths]
+input_cache = caches[0]
+palette_cache = caches[1]
+background_cache = caches[2]
+foreground_cache = caches[3]
+
+# Define a function to retrieve embeddings from the cache if present, and otherwise request via the API
+def retrieve_embedding(string: str,
+    cache_path: str,
+    embedding_cache,
+    model: str = EMBEDDING_MODEL,
+) -> list:
+    """Return embedding of given string, using a cache to avoid recomputing."""
```

Comment:
To be honest, this was mainly because I was following along with an [OpenAI walkthrough](https://cookbook.openai.com/examples/recommendation_using_embeddings) and their comment for this method happened to be in triple quotes. I mainly just overlooked this while trying to adapt their original function to taking in multiple sources of caches as input